Feature: Enable and Disable plugins from in-game

samrocketman opened the pull request:
Summary
-------

This extends the in-game Plugins UI to allow the player to enable and disable plugins.

Enabling and disabling plugins is inconvenient and requires a player to either move plugins outside of the plugin folder or completely remove the plugin folder.

Feature Details
---------------

Extends `Preferences > Plugins` UI to show check boxes next to plugins.

- A checked box means the plugin is enabled.
- An unchecked box means the plugin is disabled.

Plugins are always enabled by default.  A global resources `plugins.txt` preferences file is supported because the game supports global plugins and should also optionally disable plugins by default.

Disabling or enabling plugins requires the game to be restarted before changes take effect.

closes https://github.com/endless-sky/endless-sky/issues/7171

UI Screenshots
--------------

Plugins enabled

![image](https://user-images.githubusercontent.com/875669/188227623-8895e5bf-f008-4a7d-bc50-2f368d4b4de7.png)

Multiple plugins disabled

![image](https://user-images.githubusercontent.com/875669/193434839-b6526686-1bda-49ff-b458-068dcbe3f00e.png)


Warnings UI for when user makes plugin selection changes

![image](https://user-images.githubusercontent.com/875669/193474878-4b36b935-f6f8-4911-8382-7fd87c6d3fb9.png)

Testing Done
------------

- [x] Launch ES for first time with no `plugins.txt`.  All plugins were
  enabled by default.
- [x] Launch ES for first time with global Resources `plugins.txt`.  The plugins specified were disabled by default.  User-configured settings in-game persisted enabling and disabling plugins.
- [x] I verified user settings override global plugins.txt settings.
- [x] I verified user preferences for enabled and disabled plugins persist across restarting the game.
- [x] I verified that `about.txt` plugin information loads for disabled plugins so that the user knows what they are and can optionally enable the plugins.

Performance Impact
------------------

None

samrocketman commented:
Squashed and rebased since nobody's reviewed it yet.

Hecter94 commented:
This PR looks like it would close:
https://github.com/endless-sky/endless-sky/issues/1255

A couple of other issues related to plugin UI for reference:
https://github.com/endless-sky/endless-sky/issues/3079
https://github.com/endless-sky/endless-sky/issues/6154
https://github.com/endless-sky/endless-sky/issues/2022

tehhowch commented on source/Plugins.cpp:
```diff
@@ -0,0 +1,82 @@
+
+	void LoadSettings(const string &path)
+	{
+		DataFile prefs(path);
+		for(const DataNode &node : prefs)
+			settings[node.Token(0)] = (node.Size() == 1 || node.Value(1));
```
Let's be just a bit more future-proof:
1. Store the data under a key, rather than treat the file itself as the key.
2. Require an argument, rather than treating the token alone as enabled.

samrocketman replied:
I used `preferences.txt` as a reference.  I think I understand what you mean but I don't have examples.  I'm currently using either `preferences.txt` or `keys.txt` as references and they're not under a key.

When you say key do you mean something like this:

```
plugins
	some-plugin 1
	another-plugin 0
```

I'm not sure what you mean by require an argument.  As in skip if there's not a token or value?

tehhowch replied:
> I used `preferences.txt` as a reference.  I think I understand what you mean but I don't have examples.  I'm currently using either `preferences.txt` or `keys.txt` as references and they're not under a key.
> 
> When you say key do you mean something like this:
> 
> ```
> plugins
> 	some-plugin 1
> 	another-plugin 0
> ```

Yep. Using a key will allow adding additional data that is not exactly of the same type/purpose to the file in a non-breaking manner while also letting us avoid order-dependent logic when we do that. I know it is highly desired for there to be pilot-specific plugin configurations, so while we don't hit that here, we should try to make getting there as simple as possible.

> I'm not sure what you mean by require an argument.  As in skip if there's not a token or value?

Yep, require each plugin line to have 2 tokens.

samrocketman replied:
Ok I understand

tehhowch commented on source/PreferencesPanel.cpp:
```diff
@@ -584,26 +585,43 @@ void PreferencesPanel::DrawPlugins()
+		Point zoneOffset = checkboxPos + Point(0., table.GetRowBounds().Height() / 3);
+		Rectangle zoneBounds = Rectangle(zoneOffset, zoneDimension);
+		AddZone(zoneBounds, [&]() { Plugins::TogglePlugin(plugin.first); });
+		if(pluginEnabled)
+			table.Draw(plugin.first, isSelected ? bright : medium);
+		else
+			table.Draw(plugin.first, isSelected ? bright : back);
```
```suggestion
		if(isSelected)
			table.Draw(plugin.first, bright);
		else
			table.Draw(plugin.first, pluginEnabled ? medium : back);
```

tehhowch commented on source/main.cpp:
```diff
@@ -32,6 +32,7 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 #include "Panel.h"
 #include "PlayerInfo.h"
 #include "Preferences.h"
+#include "Plugins.h"
```
Ordering nitpick

samrocketman replied:
Will do

tehhowch commented on source/PreferencesPanel.cpp:
```diff
@@ -584,26 +585,43 @@ void PreferencesPanel::DrawPlugins()
 			table.DrawHighlight(back);
-		table.Draw(plugin.first, isSelected ? bright : medium);
+		bool pluginEnabled = Plugins::IsEnabled(plugin.first);
+		Point checkboxPos = table.GetRowBounds().TopLeft();
+		checkboxPos.X() -= font.Height() / 2;
+		checkboxPos.Y() += table.GetRowBounds().Height() / 2;
+
+		SpriteShader::Draw(box[pluginEnabled], checkboxPos);
+		Point zoneDimension = Point(20., table.GetRowBounds().Height() * 2. / 3.);
+		Point zoneOffset = checkboxPos + Point(0., table.GetRowBounds().Height() / 3);
```
Lots of calls to `GetRowBounds().Height()` suggests you can improve this

samrocketman replied:
Yeah, I was actually struggling with this clickable zone.  It's still not perfect.  I tried making everything relative to the table row.  I can DRY it up some.

tehhowch reviewed (commented):
Great first take!

However, there's no signaling about the need to restart the game after modifying the plugin list. Please add a notification in the UI that the current pending changes to the plugin list will not take effect until after the game is restarted.

Probably the simplest way is to add an interface element to the settings HUD and set its visible condition when the plugins list has been modified after loading.

There's also probably a need to reflect active plugins vs installed plugins in the player's save file.

samrocketman reviewed (commented):
@tehhowch said:

> Great first take!

Thanks :smile: 

> However, there's no signaling about the need to restart the game after modifying the plugin list. Please add a notification in the UI that the current pending changes to the plugin list will not take effect until after the game is restarted.

I agree and want this.  The original issue I filed with the mock actually mentions a warning message for the user.  I tried using `PreferencePanel::DoHelp()` from within the TogglePlugin lambda but it didn't fire.  I need to do some debugging and figure out how to do this.  I'm still not sure.

> Probably the simplest way is to add an interface element to the settings HUD and set its visible condition when the plugins list has been modified after loading.

I'm not sure what you mean by this.  I'll search through the code for the hud you're referencing.  Could you elaborate?

> There's also probably a need to reflect active plugins vs installed plugins in the player's save file.

I know the section of code this is referring.  I will add this; good feedback.

quyykk commented on source/Plugins.cpp:
```diff
@@ -0,0 +1,82 @@
+void Plugins::Load()
+{
+	// Global plugin settings
+	LoadSettings(Files::Resources() + "plugins.txt");
```
Any reason for not using `preferences.txt` to store the list of enable plugins? It seems fitting.

samrocketman replied:
No particular reason.  I would need to extend preferences to differentiate setting or plugin.  Having a plugins.txt felt cleaner because I could define default plugins settings for global plugins.  It can have its own validation separate from preferences.  plugins.txt felt like a natural item next to the plugins folder

quyykk commented on source/GameData.cpp:
```diff
@@ -830,6 +831,16 @@ void GameData::LoadSources()
 		icon->ValidateFrames();
 		spriteQueue.Add(icon);
+
+		// Enabling plugins is the default preference.
+		if(Plugins::IsEnabled(name))
+		{
+			Plugins::Set(name, true);
+			++it;
+			continue;
+		}
+		// Remove sources for user-disabled plugins while preserving about.txt in preferences.
+		it = sources.erase(it);
```
Instead of removing a source after the fact, don't add the source to the list in the first place.

samrocketman replied:
This is a catch 22.  I need to walk the sources to get the name.  I also need the about.txt to load for disabled plugins.  If this information isn't loaded then we could have a blank entry for disabled plugins.

I opted for this method because having the about.txt information for disabled plugins is user friendly.

samrocketman replied:
I'm going to keep this in since I want plugin preferences to show icon and `about.txt` for plugins.

quyykk replied:
You can separate this: Make a new function at the top of the file that simply loads a given plugin. Then, when traversing the directory looking for plugins, call that new function and then only add the plugin to the source list if it is enabled.

samrocketman replied:
Are we trying to avoid popping items off of a list?  The list is rarely above 20 plugins/items (overflows plugin UI).  Popping items out of a list is a cheap operation adjusting pointers.

Are you suggesting deleting this icon/about loop entirely in favor of a function to be called inside of the other global/local loops?

samrocketman replied:
Should it be a namespaced function or private function?

samrocketman replied:
I pushed 0bad9f0557d12aa4a71f1bea1f023bc22c9ae933 to address your feedback.  @quyykk if you're satisfied I will defer to you to resolve this feedback by clicking the button since you've already unresolved it before.

tehhowch commented:
> I agree and want this.  The original issue I filed with the mock actually mentions a warning message for the user.  I tried using `PreferencePanel::DoHelp()` from within the TogglePlugin lambda but it didn't fire.  I need to do some debugging and figure out how to do this.  I'm still not sure.

`DoHelp` / some other one-time notification would be insufficient for this IMO.

> > Probably the simplest way is to add an interface element to the settings HUD and set its visible condition when the plugins list has been modified after loading.
> 
> I'm not sure what you mean by this.  I'll search through the code for the hud you're referencing.  Could you elaborate?  I was thinking of using a namespace boolean and set it (but not persist it).

Mainly I think having some string displayed in any menu (not just the plugins menu) if plugins have changed is good enough for this PR. A follow-up PR could add an additional notification when the settings / controls / plugins menu closes that prompts to "quit now" (someday to reload) or "keep playing with {{list of originally active plugins}}"

To render this notification string, we can either tack it on to an existing `Interface` definition, or add a new `Interface` that will be used to display this kind of persistent notification. (I would probably start with an existing interface definition.) You'd use an `Information` object along with the [`visible if`](https://github.com/endless-sky/endless-sky/blob/master/data/interfaces.txt#L251) directive to only show it when the current state of plugin enabled-ness is different than the original state.
(The message should not show if a user toggles the same plugin twice.)

samrocketman commented:
> DoHelp / some other one-time notification would be insufficient for this IMO

I'll make an attempt.  It will likely be rough first time around.  In addition to showing "restart game" text in the UI I think it would be friendly to highlight changed plugin state in red text.

samrocketman commented:
About to push up a new UI for when user makes changes 
![image](https://user-images.githubusercontent.com/875669/188286675-a3914232-acd9-46d5-b0d9-23495467fde7.png)

samrocketman commented:
I'm about to push a commit addressing the last bit of feedback.  The user save file looks like the following

```
# Installed plugins:
plugins
	AES-Omnis
	ES-news-portraits
	ES_Lampyrid_Start
	Lost-in-Midnight
	Mega-Freight
	Midnight-Scrapyard
	all-content-plugin
	endless-sky-high-dpi (disabled)
```

And the `~/.local/share/endless-sky/plugins.txt` settings file now uses `plugins` as the parent key for plugins being enabled or disabled.

```
plugins
	AES-Omnis 1
	ES-news-portraits 1
	ES_Lampyrid_Start 1
	Lost-in-Midnight 1
	Mega-Freight 1
	Midnight-Scrapyard 1
	all-content-plugin 1
	endless-sky-high-dpi 0
```

samrocketman commented:
@tehhowch @quyykk I think I've addressed all of your feedback.  I updated the description of this pull request with new UI screenshots.

Fzzr commented:
This suggests a future extension to also show "missing" plugins that are referenced in a save but not present on disk. Out of scope here due to changes in various other places first.

samrocketman commented:
> This suggests a future extension to also show "missing" plugins that are referenced in a save but not present on disk. Out of scope here due to changes in various other places first. Or is it, not actually tried?

I would also like to throw a dialog warning when "Enter ship" with a pilot and plugin set is different.  Like: "Are you sure you want to load?"

quyykk commented on data/interfaces.txt:
```diff
@@ -429,6 +430,14 @@ interface "preferences"
+	visible if "show plugins changed"
+	sprite "ui/planet dialog button"
+		center 0 210
+	visible if "show plugins changed"
```
```suggestion
```

The visible if condition stays until you change it, so this one is unnecessary.

samrocketman replied:
I'll want to test

quyykk commented on source/GameData.cpp:
```diff
@@ -92,6 +93,42 @@ namespace {
 	map<const System *, map<string, int>> purchases;
+
+	// Load plugin icons and description for preferences dialog.
+	bool LoadAboutPlugin(const string &path) {
```
`{` on a newline.

quyykk commented on source/Plugins.cpp:
```diff
@@ -0,0 +1,124 @@
+			if(key == "plugins")
+				for(const DataNode &child : node)
+					if(child.Size() == 2)
+						settings[child.Token(0)] = child.Value(1);
```
If you load the settings directly into `frozenSettings` you don't need the `Freeze` function.

samrocketman replied:
I think logic would be simpler with the two maps being equivalent.  Otherwise, I'll need to use `map.find` and search for `map.end()` when doing a one to one comparison between two maps.

Is that what you mean?  If not, can you elaborate?

samrocketman replied:
Adding on, even if `plugins.txt` exist, if a user added a new plugin the entry would not exist in persisted settings the first launch.  So Freeze is still required in that case as well.

samrocketman replied:
Removed frozenSettings in the latest commit in favor of a bool pair.  The freeze is still necessary but it no longer does a map copy.

quyykk commented on source/Plugins.cpp:
```diff
@@ -0,0 +1,124 @@
+void Plugins::Load()
+{
+	// Global plugin settings
+	LoadSettings(Files::Resources() + "plugins.txt");
```
Do we need global plugin settings? One plugin preference file aught to be enough.

samrocketman replied:
Yes we do.  I already use it.  We have global plugins so supporting a global settings definition makes sense.

If you'd like an example see https://github.com/samrocketman/endless-sky/blob/mining-complete-plugins/plugins.txt

These are set upon first launch by a new player.  I think we should also allow plugins to define themselves as being disabled by default but I'll open a separate PR for that.

samrocketman replied:
This would also enable the following:

- Breaking out core game data from interface data which would allow players to disable core data for complete conversion mods.
- High DPI can be pre-packaged and disabled by default.  One of the original criticisms of prepackaging high DPI was its memory and load time requirements.
- Additional logic can be added to the game where if the resolution is above a certain point that High DPI is automatically enabled.

quyykk commented on source/Plugins.cpp:
```diff
@@ -0,0 +1,124 @@
+
+void Plugins::Save()
+{
+	if(settings.size() == 0)
```
```suggestion
	if(settings.empty())
```

quyykk commented on source/GameData.cpp:
```diff
@@ -92,6 +93,42 @@ namespace {
+		// Enabling plugins is the default preference.
+		if(Plugins::IsEnabled(name))
+		{
+			Plugins::Set(name, true);
```
If the plugin is already enabled why do we need to enable it again?

samrocketman replied:
Upon first launch it will not have an entry in the settings map.  Not being in the map is considered enabled.

samrocketman replied:
I can check if an entry exists before setting it but I think this would be more wasteful than always assuming because `map.find` will be called twice.  Not a big deal since the list is small.

quyykk commented on source/Plugins.h:
```diff
@@ -0,0 +1,35 @@
+
+
+
+class Plugins {
```
Add a comment for the class please :)

samrocketman replied:
Upon first launch it will not have an entry in the settings map.  Not being in the map is considered enabled.

samrocketman commented:
Updated UI

![image](https://user-images.githubusercontent.com/875669/188353550-157328b1-94e8-4a4d-a8b4-6cd660dfb45f.png)

tibetiroka commented on source/Plugins.h:
```diff
@@ -0,0 +1,37 @@
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
```
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```

samrocketman commented on data/interfaces.txt:
```diff
@@ -96,6 +96,7 @@ color "radar blink" 1. 1. 1. 0.
 color "radar viewport" 0. .3 0. 0.
 
 # Colors used for warnings and errors.
+color "bright error" .75 .1 .1 0.
```
Should I rename this since it is only used by plugins changed warning text?

samrocketman replied:
I renamed it

tehhowch commented on source/PreferencesPanel.cpp:
```diff
@@ -583,27 +586,49 @@ void PreferencesPanel::DrawPlugins()
 	Table table;
 	table.AddColumn(-115, {MAX_TEXT_WIDTH, Truncate::MIDDLE});
-	table.SetUnderline(-120, 120);
+	table.SetUnderline(-120, 100);
 
 	int firstY = -238;
-	table.DrawAt(Point(-130, firstY));
+	// Table is at -110 while checkbox is at -130
+	table.DrawAt(Point(-110, firstY));
```
Currently `Table` only knows how to draw text. Would it make sense to support drawing a sprite centered in the active column? This would align `Table` a bit more with HTML tables (minus the bits about automatic bounds).

samrocketman replied:
I’m not sure.  I am open to changes.  I usually only write CLI programs for this UI code was me throwing everything at a wall to see what sticks.

tehhowch commented on source/PreferencesPanel.cpp:
```diff
@@ -583,27 +586,49 @@ void PreferencesPanel::DrawPlugins()
+		Rectangle zoneBounds = Rectangle(zoneOffset, zoneDimension);
+		AddZone(zoneBounds, [&]() { Plugins::TogglePlugin(plugin.first); });
+		if(Plugins::HasChanged(plugin.first))
+			table.Draw(plugin.first, error);
```
Definitely don't use an error color when the plugin is free of errors. If you want to reflect that the plugin is currently disabled, or still currently active, use the colors that are normally used in that scenario. The checkbox encapsulates the change that will take effect.

samrocketman replied:
The color can be renamed to “plugins changed” instead of error if you like.

There’s three colors at play.

1. Standard enabled (checked/bright)
2. Standard disabled (unchecked/dim)
3. A selection has changed from current runtime requiring restart (currently red text).

If a user changes plugin runtime the code highlights which plugins were changed and show a restart message.  If user decides to revert their selection the restart message goes away because the selection matches the runtime.  Currently, there’s no standard way of displaying this.

I can rename the variable and keep the same color or suggest a color and we can see how it looks.

Might be helpful to understand intent by building and messing with the UI.

tehhowch replied:
I'm saying that the color of the plugin name should only depend on the enabled-ness of the plugin at the time the game was started, as that is the enabled-ness of the plugin if the player enters their ship after leaving this menu.

The plugin checkbox state reflects the enabled-ness following the next restart.

samrocketman replied:
That was how the original PR worked.  Since making the changes with the red text (I am flexible on the color).  It is convenient if I or another user changes our mind and decides to not toggle plugins being enabled.

By reverting selections to match startup the need to restart goes away.

For a large number of plugins (greater than 10) one does not need to memorize their previous selection.

tehhowch replied:
Why do you mention memorizing? The scheme I propose requires no memorization by the user:
1) draw the plugin name as either `bright` or `dim`, based on the initial state of the plugin, when the game booted.
2) draw the checkbox as selected or unselected, based on the current state of the plugin.

Using an error color when the plugin has no errors in it is misleading. The game does not _currently_ evaluate plugin definitions for errors, but it _will_.

samrocketman replied:
I don't think you understand the current implementation just by reading the code.  You need to launch the game and see how it behaves because it already works as you describe.  The warning state is when the user makes enable/disable changes.  It needs to be highlighted in case the user changes their mind.

I encourage you to:
1. launch the game.
2. Disable a few plugins (not all)
3. Restart  the game.  You will see dim and normal text based on current state.
4. Make some changes and you will see how they highlight changes you've made from the current runtime.

Memory is definitely an issue when you have state tracked for a dozen plus plugins and then change their runtime for next restart.  The coloring solves the problem user would have to remember their selection if they want to undo their changes.

samrocketman replied:
I renamed the code reference from `error` to `pluginReloadRequired`.  This is hopefully more clear as to the color's purpose and not to be an error.

samrocketman replied:
![image](https://user-images.githubusercontent.com/875669/193474878-4b36b935-f6f8-4911-8382-7fd87c6d3fb9.png)

Here's a screenshot of your suggestion.  The current selected plugin is hard to tell.  You have to move your mouse a few times to get the full view of what plugins changed in terms of user clicking check boxes.

samrocketman replied:
Pushed the suggestion, though I think from a usability perspective it's more difficult to change plugin selections without some kind of button to revert back to initial selection.  e.g. restore defaults.

tehhowch commented on source/PreferencesPanel.cpp:
```diff
@@ -583,27 +586,49 @@ void PreferencesPanel::DrawPlugins()
+		checkboxPos.Y() += rowHeight / 2;
+
+		SpriteShader::Draw(box[pluginEnabled], checkboxPos);
+		Point zoneDimension = Point(20., thirdHeight * 2.);
+		Point zoneOffset = checkboxPos + Point(0., thirdHeight - 3);
+		Rectangle zoneBounds = Rectangle(zoneOffset, zoneDimension);
```
It might make sense to add a Rectangle constructor that takes an existing rectangle and then the amount to expand or contract its vertical and horizontal dimensions.

samrocketman replied:
This was trial and error guessing.  Both for code and dimensions.  I have no meaningful way to adjust the UI code other than changing values and building.

I will need a bit more hand holding on this request because I don’t understand and am new to graphics code.

tehhowch replied:
My comment was more of a "geometric operations simplification" request, since you're creating a box from another box, with a bunch of intermediate steps.

samrocketman replied:
I couldn't seem to get the boxes to line up initially and that's what all the extra math is for.  It was trial and error because one appears to use the center while the other uses some other dimension I couldn't figure out because it was invisible (maybe a corner?).

The current sizing was me trying to work within those constraints while not quite understanding why the two shapes didn't line up originally.

samrocketman replied:
If we do this I think it should be a separate PR.

samrocketman commented on source/Plugins.cpp:
```diff
@@ -0,0 +1,133 @@
+// Freeze the plugin set to determine if a setting has changed.
+void Plugins::Freeze()
+{
+	copy(settings.begin(), settings.end(), inserter(frozenSettings, frozenSettings.end()));
```
I’m not sure I understand; do you mean with a loop and do

```
for(…)
{
    frozenSettings[key] = settingsValue;
}
```

As a pseudo example?

samrocketman replied:
I created a copy because I didn't want changes in settings to also affect frozenSettings.  I can do some debugging to verify if `frozenSettings = settings` is good enough.  Would be good if @tehhowch can confirm that's what they mean.

tehhowch replied:
That's exactly what I mean.

samrocketman replied:
Ok I will try with const as well.  It sounds easy

samrocketman replied:
The implementation has changed and copying the map is no longer used.  Instead a boolean pair.  Please see the latest diff for `Plugins::Freeze()` method.  Does this work to resolve feedback?

samrocketman commented on source/Plugins.cpp:
```diff
@@ -0,0 +1,133 @@
+		return;
+	DataWriter out(Files::Config() + "plugins.txt");
+
+	out.Write("plugins");
```
As in rename the keyword to state?  Kind of tricky what to call it because as part of future proofing it can be assumed we’ll eventually expose plugin settings UI to users.

I think the name has to give meaning in the context of other future-existing settings.

`state` isn’t descriptive and I agree `plugins` isn’t descriptive/slightly redundant.  Not sure what to call this one.

samrocketman replied:
The save file lists `plugins` which is why I matched it.

samrocketman replied:
It is now state

samrocketman commented on source/Plugins.cpp:
```diff
@@ -0,0 +1,133 @@
+// Determine if a plugin setting has changed since launching.
+bool Plugins::HasChanged(const string &name)
+{
+	return frozenSettings[name] != settings[name];
+}
+
+
+
+// Returns true if any plugin enabled or disabled setting has changed since
+// launched via user preferences.
+bool Plugins::HasChanged()
+{
+	for(const auto &plugin : frozenSettings)
+		if(plugin.second != settings[plugin.first])
```
I don’t understand.  Can you elaborate and guide me toward what is good in this case?  Overall purpose of this function is to compare equality of two maps.  So that “restart game” text is displayed to user in pause menu.

Hurleveur replied:
I think he means you need to use a method to get what the data of this key is, explicitely, one that would be const qualified (and thus wouldnt require a const conversion?)
I dont have a computer rn so idk what method that is

samrocketman replied:
I'm still not sure I understand the terminology.  I web searched in order to learn about it but the results were not something I could learn to address this feedback.  So an example of some kind would be great.

tehhowch replied:
This method name and signature strongly imply that no changes will be caused by simply invoking it.

However, it uses `operator[]`, which modifies the subject map and inserts a default-constructed value of the appropriate type. If you have a `const map<K, V> foo`, you cannot compile `foo[someKey]` as `foo` is not modifiable.

A const-qualified inclusion check would generally use `map::find`.

samrocketman replied:
Thanks for expanding.  I will try to use assignment with the variable being const type.

samrocketman replied:
@warp-core helped me through a better const-qualified solution I think.  See b80a33425ba706ee1d12b3bc14dd26935f575a35 which removes frozenSettings and instead uses a pair of bool values.  `map.find` is used instead of `operator[]` accessor.

samrocketman commented on data/interfaces.txt:
```diff
@@ -429,6 +430,17 @@ interface "preferences"
+	visible if "show plugins changed"
+	sprite "ui/planet dialog button"
+		center 0 210
+	label "Restart game"
+		center 0 201
+		dimensions 140 40
+		color "bright error"
+	label "to reload plugins."
+		center 0 219
+		dimensions 140 40
+		color "bright error"
```
It looks less like a button with two levels of text.  I can create a UI graphic similar to the other UI panels if that’s desirable.

Without the panel/border the text is not easily read (practically impossible for poor eye sight and really straining on my good eyesight).

samrocketman replied:
@Terin has provided a good suggestion for this.  `Installed plugins:` now changes to `Restart to apply changes`.  Here's a screenshot. I will update the pull request description as well.

![image](https://user-images.githubusercontent.com/875669/193434471-7ac86715-d1b2-4759-acd3-1afe2b7b1534.png)

samrocketman commented:
@Saugia I don't think this is waiting on me.  I've addressed all feedback and asked some followup questions which haven't been answered.  So there's not much for me to do here.

quyykk reviewed (commented):
This is really nice!

-  The disabled plugins are too dark, it should be the "dim" color I think to follow the convention used by the rest of the game.

I'm not totally sure about the red text, since it would be a first. How about a light red background, similar to the yellow background drawn for conflicting control keys?

I'm not sure what we could use instead of the red button though.

warp-core commented on source/GameData.cpp:
```diff
@@ -797,43 +836,21 @@ void GameData::LoadSources()
 	for(const string &path : localPlugins)
 	{
 		if(Files::Exists(path + "data") || Files::Exists(path + "images") || Files::Exists(path + "sounds"))
-			sources.push_back(path);
+			if(LoadAboutPlugin(path))
+				sources.push_back(path);
```
```
	if((Files::Exists(path + "data") || Files::Exists(path + "images") || Files::Exists(path + "sounds"))
			&& LoadAboutPlugin(path))
		sources.push_back(path);
```
?

samrocketman replied:
I made this change in two places and attributed you.

kestrel1110 commented:
So does this make it so you can have custom presets or "loadouts" of plugins per each pilot, or simply make a global interface that effects all pilots?

warp-core commented on source/GameData.cpp:
```diff
@@ -797,43 +836,21 @@ void GameData::LoadSources()
-		icon->ValidateFrames();
-		spriteQueue.Add(icon);
-	}
+	// Freeze the current plugin state to check if plugin state has changed when
+	// toggling user settings.  This should only be called once per instance.
+	Plugins::Freeze();
```
I don't think you need this anymore?

samrocketman replied:
This has been deleted since implementing tehhowch's struct suggestion.

warp-core commented on source/PlayerInfo.cpp:
```diff
@@ -3138,7 +3139,12 @@ void PlayerInfo::Save(const string &path) const
-			out.Write(plugin.first);
+		{
+			if(Plugins::IsEnabled(plugin.first))
+				out.Write(plugin.first);
```
I think this should include the state.
```suggestion
				out.Write(plugin.first, "(enabled)");
```

samrocketman replied:
I can add this.  I implied if it didn't have disabled that it was enabled but we can make it more explicit.

samrocketman replied:
I added this and attributed you.

warp-core commented on source/Plugins.h:
```diff
@@ -0,0 +1,40 @@
+
+#include <string>
+
+
```
```suggestion


```

samrocketman replied:
Added

warp-core commented on source/Plugins.h:
```diff
@@ -0,0 +1,40 @@
+	static bool HasChanged();
+	static bool HasChanged(const std::string &name);
+	static bool IsEnabled(const std::string &name);
+	static void Set(const std::string &name, bool on);
```
```suggestion
	static void SetPlugin(const std::string &name, bool on);
```
To match `TogglePlugin`.

samrocketman replied:
Renamed to `SetPlugin(const std::string &name, bool on)`.  Also added an overloaded method `SetPlugin(const std::string &name)` which is used during GameData loading of `plugins.txt`.

warp-core commented on source/Plugins.cpp:
```diff
@@ -0,0 +1,135 @@
+			if(key == "plugins")
+				for(const DataNode &child : node)
+					if(child.Size() == 2)
+						settings[child.Token(0)].first = child.Value(1);
```
Any reason not to set both bools here?
```suggestion
						settings[child.Token(0)] = make_pair(child.Value(1), child.Value(1));
```
I'm not sure if there is a performance penalty to calling `child.Value(1)` twice, the compiler might just cache it (unless it doesn't know that the return value won't change, I suppose that depends on if `child.Value()` is `const`).
```suggestion
					{
						bool state = child.Value(1);
						settings[child.Token(0)] = make_pair(state, state);
					}
```

samrocketman replied:
Resolved by using tehhowch struct with support for setting the initial state via constructor.

samrocketman commented:
> So does this make it so you can have custom presets or "loadouts" of plugins per each pilot, or simply make a global interface that effects all pilots?

Global menu to enable or disable plugins in-game.

Currently dynamic plugin loading is not supported

warp-core commented on source/Plugins.cpp:
```diff
@@ -0,0 +1,135 @@
+
+
+
+// Freeze the plugin set to determine if a setting has changed.
+void Plugins::Freeze()
+{
+	for(auto &it : settings)
+		it.second.second = it.second.first;
+}
+
+
+
```
Don't think we need this anymore.
```suggestion
```

samrocketman replied:
We 100% need this.  See what I responded to quyykk on why this is necessary.  We need to capture plugins which don't exist in plugins.txt.

samrocketman replied:
This has been deleted; thanks.

warp-core commented on source/Plugins.cpp:
```diff
@@ -0,0 +1,135 @@
+
+
+
+// Plugins are enabled by default and disabled if the user prefers it to be
+// disabled.
+bool Plugins::IsEnabled(const string &name)
+{
+	auto it = settings.find(name);
+	return (it == settings.end()) || it->second.second;
+}
```
Is this `enabled` as in, loaded, or as in the player has set it to enabled (but it might not be loaded because it was previously disabled, they've enabled it in the UI, but haven't restarted yet)?

samrocketman replied:
IsEnabled is the player setting because it is used as part of writing config.  If you want to determine if if the state is different than loaded runtime use HasChanged method.

samrocketman replied:
HasChanged has been removed.  How there's `InitialPluginState` method to get the state from when the game first started.

warp-core commented on source/Plugins.cpp:
```diff
@@ -0,0 +1,135 @@
+
+
+
+void Plugins::Set(const string &name, bool on)
```
```suggestion
void Plugins::SetPlugin(const string &name, bool on)
```

samrocketman replied:
Renamed

tehhowch commented on source/Plugins.cpp:
```diff
@@ -0,0 +1,135 @@
+
+namespace {
+	// The first item in pair is plugin state on load, second item is user changing preference.
+	map<string, pair<bool, bool>> settings;
```
You can use an anonymous class type if you would rather not deal with `.second.first` and `.second.second`.

E.g.
```c++
struct EnabledState {
  bool initial = true;
  bool current = true;
}
map<string, EnabledState> settings;
```

samrocketman replied:
Can you elaborate how this would be used?  I see the definition but not clear to me how I would instantiate it or use it.

samrocketman replied:
I added the `EnabledState` struct.  I figured out how to use it once I was in an IDE which gave me hints on how to use it.

I also added a constructor which allows me to customize the `initial` and `current` state which gets used on GameData loading of `plugins.txt`.

tehhowch commented on source/Plugins.cpp:
```diff
@@ -0,0 +1,135 @@
+bool Plugins::HasChanged(const string &name)
+{
+	const auto it = settings.find(name);
+	return (it == settings.end()) || it->second.first != it->second.second;
```
Shouldn't this return false when the value is not in the map? (The plugin has definitely not changed state if we didn't know about it.)

```suggestion
	return it == settings.end() ? false :  it->second.first != it->second.second;
```

warp-core replied:
The idea was, if the plugin isn't in the map, something has gone wrong and you should probably restart anyway.

samrocketman replied:
This is more of an edge case.  In reality it should be impossible to ever reach `settings.end()` so at best reaching that point should he some kind of error.  For now, the way the error is surfaced for this case is to warn the user to restart.

samrocketman replied:
HasChanged == true is intentional when `settings.end()` is reached.  I want to keep `settings.end()` protection to prevent crashes.  But in general it shouldn't ever return `true`.

Hurleveur reviewed (approved):
Just looked at the code and I couldnt find any issues.
This could be improved further but that'd be best left for a further PR.

samrocketman commented:
There might be a bug.  I’ll investigate and report back.

samrocketman commented:
I have verified there is no bug.

Hurleveur commented:
Did you see that in the issue smbd asks for the default to be plugins being offline?

quyykk commented on source/Plugins.cpp:
```diff
@@ -0,0 +1,148 @@
+// Check if a plugin is known. It does not matter if it is enabled or disabled.
+bool Plugins::Has(const string &name)
+{
+	auto it = settings.find(name);
+	return it != settings.end();
```
```suggestion
	return settings.count(name);
```

quyykk replied:
There is no functional difference yes. Since we don't need the actual iterator, `find` is a bit unnecessary, and `count` better shows the intent. imo

quyykk commented on source/GameData.cpp:
```diff
@@ -796,44 +835,18 @@ void GameData::LoadSources()
 	vector<string> globalPlugins = Files::ListDirectories(Files::Resources() + "plugins/");
 	for(const string &path : globalPlugins)
 	{
-		if(Files::Exists(path + "data") || Files::Exists(path + "images") || Files::Exists(path + "sounds"))
+		if((Files::Exists(path + "data") || Files::Exists(path + "images") || Files::Exists(path + "sounds"))
+				&& LoadAboutPlugin(path))
```
That should also help to remove a lot of functions from `Plugins`.

Saugia commented:
I was trying to see if I can delete the comment you mentioned my name in so it no longer mentions me, but it seems that stays. Oh well.

Hurleveur commented:
![image](https://user-images.githubusercontent.com/94366726/203920834-13802d06-5aa4-46b2-a1ea-b1ea4055db31.png)
This is what you need to change to not longer be notified

samrocketman commented:
Why was my comment deleted?  What was the comment?  Let’s avoid deleting comments in the future unless there’s a real need to (e.g. moderation of inappropriateness)

quyykk commented:
Definitely agree with Sam there. Also on another note, Sam, did you see the PR that I opened to your branch. Might also be interesting to @tehhowch. It basically addresses every outstanding review comment to the beat of my knowledge (except for merge conflicts).

samrocketman commented:
> Definitely agree with Sam there. Also on another note, Sam, did you see the PR that I opened to your branch. Might also be interesting to @tehhowch. It basically addresses every outstanding review comment to the best of my knowledge (except for merge conflicts).

Thanks for pointing it out I hadn’t seen it since I’ve been focused in other areas of development.  I’ll take a look and can probably spend a little time to address conflicts.

RisingLeaf commented on source/Plugins.h:
```diff
@@ -0,0 +1,70 @@
+
+
+// Represents information about a single plugin.
+struct Plugin {
```
Can we make this a class in a separate file?
That would be more how we generally structure our things.

samrocketman replied:
Not sure it makes sense to break this out.  Its intent is to track state and the only place it would ever be used is this Plugins file.  Breaking out into classes seems like it would be better intended if code is meant to for reuse at some point.  I can't imagine this getting reused.

RisingLeaf replied:
so in that case, could you move this struct into the Plugins class, to match how we normally do it

samrocketman replied:
@RisingLeaf I can still open a follow up PR if you want

RisingLeaf replied:
It seems like warp core thought this is fine, so I do not think a follow up is necessary

warp-core replied:
I was going to change it but I forgot by the time I got to the end of the changes and tested it.

RisingLeaf reviewed (changes requested):
Except for this small thing, this looks really good.

quyykk commented:
@tehhowch do you still want to review this?
